### PR TITLE
ER-470 Reporting command

### PR DESCRIPTION
**Pull request by chrisjensenuk**
Next batch of changes.  This is to do with the create report command.

Fluent validation defined in the various web projects now need to be explicitly registered in IOC as I have removed the RegisterValidatorsFromAssemblyContaining function call.

- Created Reports collection
- Added CreateReportCommand
- Added CreateReportCommandHandler
- Adds a report to the Reports collection

### CODE REVIEW

File: src/Provider/Provider.Web/ViewModels/Reports/ProviderApplicationsReport/ProviderApplicationsReportCreateEditModel.cs [hunk 1]
```diff
@@ -1,7 +1,12 @@
-﻿namespace Esfa.Recruit.Provider.Web.ViewModels.Reports.ProviderApplicationsReport
+﻿using Microsoft.AspNetCore.Mvc;
+
+namespace Esfa.Recruit.Provider.Web.ViewModels.Reports.ProviderApplicationsReport
 {
     public class ProviderApplicationsReportCreateEditModel
     {
+        [FromRoute]
```
Comment (dev4valtech) [inline src/Provider/Provider.Web/ViewModels/Reports/ProviderApplicationsReport/ProviderApplicationsReportCreateEditModel.cs]: This should be derived from User claims

File: src/Shared/Recruit.Vacancies.Client/Domain/Events/ReportCreatedEvent.cs [hunk 2]
```diff
@@ -0,0 +1,11 @@
+﻿using System;
+using Esfa.Recruit.Vacancies.Client.Domain.Messaging;
+using MediatR;
+
+namespace Esfa.Recruit.Vacancies.Client.Domain.Events
+{
+    public class ReportCreatedEvent : EventBase, INotification
```
Comment (dev4valtech) [inline src/Shared/Recruit.Vacancies.Client/Domain/Events/ReportCreatedEvent.cs]: What is the purpose of this event?

File: src/Employer/Employer.Web/Configuration/ConfigurationExtensions.cs [hunk 3]
```diff
@@ -88,8 +88,7 @@ public static void AddMvcService(this IServiceCollection services, IHostingEnvir
                 opts.Filters.AddService<PlannedOutageResultFilter>();
 
                 opts.AddTrimModelBinderProvider(loggerFactory);
-            })
-            .AddFluentValidation(fv =>fv.RegisterValidatorsFromAssemblyContaining<Startup>())  
+            })
```
Comment (shomavg) [inline src/Employer/Employer.Web/Configuration/ConfigurationExtensions.cs]: Can you please test if the validations are still working on application reviews?

File: src/Shared/Recruit.Vacancies.Client/Application/CommandHandlers/CreateReportCommandHandler.cs [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+    {
+        private readonly ILogger<CreateReportCommandHandler> _logger;
+        private readonly IReportRepository _repository;
+        private readonly IMessaging _messaging;
+        private readonly ITimeProvider _timeProvider;
+
+        public CreateReportCommandHandler(
+            ILogger<CreateReportCommandHandler> logger,
+            IReportRepository repository, 
+            IMessaging messaging,
```
Comment (shomavg) [inline src/Shared/Recruit.Vacancies.Client/Application/CommandHandlers/CreateReportCommandHandler.cs]: In there for future use?

File: src/Shared/Recruit.Vacancies.Client/Domain/Events/ReportCreatedEvent.cs (hunk 2 above)
Comment (chrisjensenuk) [reply src/Shared/Recruit.Vacancies.Client/Domain/Events/ReportCreatedEvent.cs]: 👍 leftovers. Will remove

File: src/Provider/Provider.Web/ViewModels/Reports/ProviderApplicationsReport/ProviderApplicationsReportCreateEditModel.cs (hunk 1 above)
Comment (chrisjensenuk) [reply src/Provider/Provider.Web/ViewModels/Reports/ProviderApplicationsReport/ProviderApplicationsReportCreateEditModel.cs]: I'm happy with this. In future we could have a god user that can navigate between providers so we should let the route dictate which provider the user is targeting. So I think decoupling resource identification from authorisation is a good thing.

The authorisation filter is triggered on every page request so I know I can trust this ukprn.

File: src/Employer/Employer.Web/Configuration/ConfigurationExtensions.cs (hunk 3 above)
Comment (chrisjensenuk) [reply src/Employer/Employer.Web/Configuration/ConfigurationExtensions.cs]: 👍 Will do

Comment (chrisjensenuk) [reply src/Employer/Employer.Web/Configuration/ConfigurationExtensions.cs]: Doh! Still have to configure FluentValidation for MVC. Have now fixed. Good shout!

File: src/Shared/Recruit.Vacancies.Client/Application/CommandHandlers/CreateReportCommandHandler.cs (hunk 4 above)
Comment (chrisjensenuk) [reply src/Shared/Recruit.Vacancies.Client/Application/CommandHandlers/CreateReportCommandHandler.cs]: That was the plan but in the end I'm not using messaging so I have now removed.

File: src/Provider/Provider.Web/ViewModels/Reports/ProviderApplicationsReport/ProviderApplicationsReportCreateEditModel.cs (hunk 1 above)
Comment (dev4valtech) [reply src/Provider/Provider.Web/ViewModels/Reports/ProviderApplicationsReport/ProviderApplicationsReportCreateEditModel.cs]: While I am not sure what the "god" user requirements are, I am happy that authorisation filter is safe guarding the value in URL.

**Pull request merged**